sdks/java/core/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataInboundObserverTest.java

```diff
@@ -165,6 +171,11 @@ public void testCloseVisibleToAwaitCompletionCallerAndProducer() throws Exceptio
     Future<?> future2 =
         executor.submit(
             () -> {
+              synchronized (isReady) {
```

Comment:
sorry, actually this cannot be synchronized (either using this or isReady). Because synchronized block will block other thread's same synchronized block. Here it will wait for signal indefinitely and cause deadlock.

Reply:
Using synchronized block on isReady works here. I don't see there is a deadlock.

(1) If it runs the first synchronized block first, then it will call notify() then isReady is set to true. Then when the second synchronized block is executed, it will not reach the wait() call because isReady is true.

(2) If it runs the second synchronized block first, then it will wait() and give up the lock, while in the first synchronized block, it will pick up the lock, set isReady to true, and then call notify().

Reply:
You are right, I confused myself, LGTM